packages/axelar-soroban-std/src/shared_interfaces.rs

```diff
@@ -0,0 +1,398 @@
+        });
+
+        let client = UpgradeableClient::new(&env, &contract_id);
+        assert_invoke_auth_ok!(owner, client.try_upgrade(&hash));
+    }
+
+    #[test]
+    fn migrate_fails_if_caller_not_authenticated() {
+        let env = Env::default();
+        let contract_id = env.register(testdata::contract::Contract, ());
+        let hash = env.deployer().upload_contract_wasm(WASM);
+
+        let owner = Address::generate(&env);
+        env.as_contract(&contract_id, || {
+            shared_interfaces::set_owner(&env, &owner);
+        });
+
+        let upgrade_client = UpgradeableClient::new(&env, &contract_id);
+        assert_invoke_auth_ok!(owner, upgrade_client.try_upgrade(&hash));
+
+        let client = ContractClient::new(&env, &contract_id);
+        assert!(client.try_migrate(&()).is_err());
+    }
+
+    #[test]
+    fn migrate_fails_if_called_by_non_owner() {
+        let env = Env::default();
+        let contract_id = env.register(testdata::contract::Contract, ());
+        let hash = env.deployer().upload_contract_wasm(WASM);
+
+        let owner = Address::generate(&env);
+        env.as_contract(&contract_id, || {
+            shared_interfaces::set_owner(&env, &owner);
+        });
+
+        let upgrade_client = UpgradeableClient::new(&env, &contract_id);
+        assert_invoke_auth_ok!(owner, upgrade_client.try_upgrade(&hash));
+
+        let client = ContractClient::new(&env, &contract_id);
+        assert_invoke_auth_err!(Address::generate(&env), client.try_migrate(&()));
+    }
+
+    #[test]
+    fn migrate_fails_if_not_called_after_upgrade() {
+        let env = Env::default();
+        let contract_id = env.register(testdata::contract::Contract, ());
+
+        let owner = Address::generate(&env);
+        env.as_contract(&contract_id, || {
+            shared_interfaces::set_owner(&env, &owner);
+        });
+
+        let client = ContractClient::new(&env, &contract_id);
+        assert_invoke_auth_err!(owner, client.try_migrate(&()));
+    }
+
+    #[test]
+    fn migrate_succeeds_if_owner_is_authenticated_and_called_after_upgrade() {
+        let env = Env::default();
+        let contract_id = env.register(testdata::contract::Contract, ());
+        let hash = env.deployer().upload_contract_wasm(WASM);
+
+        let owner = Address::generate(&env);
+        env.as_contract(&contract_id, || {
+            shared_interfaces::set_owner(&env, &owner);
+        });
+
+        let upgradeable_client = UpgradeableClient::new(&env, &contract_id);
+        assert_invoke_auth_ok!(owner, upgradeable_client.try_upgrade(&hash));
+
+        let client = ContractClient::new(&env, &contract_id);
+        assert!(client.migration_data().is_none());
+
+        assert_invoke_auth_ok!(owner, client.try_migrate(&()));
+
+        assert_eq!(
+            client.migration_data(),
+            Some(String::from_str(&env, "migrated"))
+        );
+
+        let event = env
```

Comment:
This is a nice pattern over the current usage of `assert_last_emitted_event` to test events. Can you create a helper that converts the event emitted at some index to the arbitrary event type `T`, could add the last_* version of this function as well, since it's commonly used. I like specifying the index better than filtering over all events, so we know if the order changes as well